### PR TITLE
http2: updates CodecImpl to use a more abstract data type than `nghttp2_settings` internally

### PR DESCRIPTION
**Pull request by birenroy**
This is refactoring in preparation for removing the nghttp2_callback API layer.

Commit Message: http2: updates CodecImpl to use a more abstract data type than `nghttp2_settings` internally
Additional Description:
Risk Level: low; refactoring, no functional change intended
Testing: ran unit and integration tests locally
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

File: source/common/http/http2/codec_impl.h [hunk 1]
```diff
@@ -596,7 +597,7 @@ class ConnectionImpl : public virtual Connection,
   void sendSettingsHelperOld(const envoy::config::core::v3::Http2ProtocolOptions& http2_options,
                              bool disable_push);
   // Callback triggered when the peer's SETTINGS frame is received.
-  virtual void onSettings(const nghttp2_settings& settings) {
+  virtual void onSettings(const std::vector<http2::adapter::Http2Setting>& settings) {
```
Comment (diannahu) [inline source/common/http/http2/codec_impl.h:600]: super optional, here and elsewhere: Could do absl::Span.

**Review by diannahu — APPROVED** [review]
Nice clean-up/readability improvement!

File: source/common/http/http2/codec_impl.h (hunk 1 above)
Comment (birenroy) [reply source/common/http/http2/codec_impl.h]: Done, thank you.

**Review by diannahu — APPROVED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

File: source/common/http/http2/codec_impl.h [hunk 2]
```diff
@@ -50,7 +51,7 @@ constexpr uint64_t H2_FRAME_HEADER_SIZE = 9;
 
 class ReceivedSettingsImpl : public ReceivedSettings {
 public:
-  explicit ReceivedSettingsImpl(const nghttp2_settings& settings);
+  explicit ReceivedSettingsImpl(const std::vector<http2::adapter::Http2Setting>& settings);
```
[the remark this answers is not in the document]
Comment (birenroy) [reply source/common/http/http2/codec_impl.h]: Yes, please see the latest commit.

[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply source/common/http/http2/codec_impl.h]: Blergh! Did I forget to hit send 2 hours ago? *sob*

**Comment by birenroy** [comment]
Ready for merge!

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**